File: 4 семестр/ПиАА/Лабораторные/PiAA_1304/Yarusova_Tatyana_lb2/src/a_star.py
import queue 

class Graph:
    # Инициализация класса
    def __init__(self):
        self.start_node = None
        self.end_node = None
        self.nodes = dict()
        self.path = ''

    # Метод, в котором происходит считывание входных данных
    def read(self):
        self.start_node, self.end_node = input().split()
        while True:
            try:
                from_node, in_node, weight = input().split()
                if from_node in self.nodes:
                    self.nodes[from_node].append([in_node, float(weight)])
                else:
                    self.nodes[from_node] = [[in_node, float(weight)]]
            except:
                break

    # Метод, реализующий эвристическую функцию
    # На вход принимает следующую вершины, в которую можно перейти
    # Возвращает разницу между кодами символов финальной вершины и вершины, поступившей на вход
    def heuristics(self, node):
        return abs(ord(node) - ord(self.end_node))
    
    # Метод, который возвращает список кортежей примыкающих вершин к текущей
    def get_neighbors(self, node):
        return self.nodes[node]
    
    # Метод, который строит и выводит путь, найденный алгоритмом A*
    # На вход получает словарь, составленный по принципу откуда пришли к текущей вершине
    def build_path(self, came_from):
        current_node = self.end_node
        while current_node != None:
            self.path += current_node
            current_node = came_from[current_node]
        self.path = self.path[::-1]
        print(self.path)


    # Метод, реализующий алгорит A*
    def a_star(self):
        nodes_queue = queue.PriorityQueue()
        nodes_queue.put((0, self.start_node))
        
        came_from = {}
        came_from[self.start_node] = None
    
        cost_so_far = {}
        cost_so_far[self.start_node] = 0

        while not nodes_queue.empty():
            current_node = nodes_queue.get()[1]

            if current_node == self.end_node:
                break
            
            if current_node in self.nodes.keys():
                neighbors = self.get_neighbors(current_node)
                for next_node, cost_next_node in neighbors:
                    new_cost = cost_so_far[current_node] + cost_next_node
                    if next_node not in cost_so_far or new_cost < cost_so_far[next_node]:
                        cost_so_far[next_node] = new_cost
                        priority = new_cost + self.heuristics(next_node)
                        nodes_queue.put((priority, next_node))
                        came_from[next_node] = current_node

        self.build_path(came_from)

# Функция, в которой создается экземпляр класса Graph
# и вызываются методы, которые реализуют поиск кратчайшего пути алгоритмом A*
def solve():
    graph = Graph()
    graph.read()
    graph.a_star()

solve()

File: 4 семестр/Построение и анализ алгоритмов/Лабораторные/PiAA_1304/Starodubov_Maxim_lb2/src/a_star.cpp
#include <iostream>
#include <functional>
#include <vector>
#include <queue>
#include <map>
#include <string>
#include <cmath>


/*
 * A priority queue is used to traverse the graph,
 * at the top is the element with the smallest path estimate f = g + h,
 * where g -- actual distance traveled to reach a given node,
 * h -- heuristic evaluation, equal to the difference in the numerical
 * values of the symbols denoting the nodes.
 * Queue elements are tuples consisting of the name of the node, the distance traveled to this node,
 * the name of the node from which the transition to this node was made.
 */
typedef std::priority_queue<std::tuple<char, float, char>, std::vector<std::tuple<char, float, char>>,
        std::function<bool(std::tuple<char, float, char> &, std::tuple<char, float, char> &)>> queue_t;
/*
 * Stores pairs of elements -- the name of the node and the weight of the edge to the this node.
 */
typedef std::vector<std::pair<char, float>> nodes_container;
/*
 * The graph is represented as an associative array, the keys are the names of the nodes, the values are the
 * containers of the nodes in which you can go to from key node.
 */
typedef std::unordered_map<char, nodes_container> graph_t;


/*
 * Class Solver is a class that encapsulates the methods needed to solve the problem.
 */
class Solution
{
public:
    graph_t get_graph();
    std::string find_path(char start_node, char finish_node, graph_t &graph);
private:
    std::string present_result(char finish_node, std::unordered_map<char, char> &path);
};


/*
 * Reads a graph from the input stream and returns it.
 */
graph_t Solution::get_graph()
{
    graph_t graph;
    char source_node, destination_node;
    float weight;

    while (std::cin >> source_node >> destination_node >> weight)
    {
        if (graph.find(source_node) == graph.end())
        {
            graph[source_node] = nodes_container();
        }
        graph[source_node].emplace_back(destination_node, weight);
    }

    return graph;
}

/*
 * Finds a path using A* algorithm.
 */
std::string Solution::find_path(char start_node, char finish_node, graph_t &graph)
{
    auto cmp = [&finish_node](std::tuple<char, float, char> &left, std::tuple<char, float, char> &right)
    {
        float left_heuristics = std::get<1>(left) + (float) std::abs(finish_node - std::get<0>(left));
        float right_heuristics = std::get<1>(right) + (float) std::abs(finish_node - std::get<0>(right));
        return left_heuristics > right_heuristics;
    };
    queue_t queue(cmp);
    std::unordered_map<char, char> path;
    std::unordered_map<char, float> best_distance;

    queue.emplace(start_node, 0, 0);

    while (!queue.empty())
    {
        auto current_item = queue.top();
        char current_node = std::get<0>(current_item);
        float distance = std::get<1>(current_item);
        char previous_node = std::get<2>(current_item);
        queue.pop();

        if (path.find(current_node) != path.end() && best_distance[current_node] <= distance)
        {
            continue;
        }

        path[current_node] = previous_node;
        best_distance[current_node] = distance;

        if (current_node == finish_node)
        {
            break;
        }

        for (auto neighbor_node: graph[current_node])
        {
            char next_node = neighbor_node.first;
            float next_node_distance = distance + neighbor_node.second;
            queue.emplace(next_node, next_node_distance, current_node);
        }

    }

    return present_result(finish_node, path);
}

/*
 * Represents the path to the finish node as a string, using an associative array,
 * the keys are the names of nodes, values are the nodes from which the transition to key node was made.
 */
std::string Solution::present_result(char finish_node, std::unordered_map<char, char> &path)
{
    std::string result;
    char current_node = finish_node;

    while (current_node)
    {
        result.insert(0, 1, current_node);
        current_node = path[current_node];
    }

    return result;
}


int main()
{
    char start_node, finish_node;
    std::cin >> start_node >> finish_node;

    Solution s;
    graph_t graph = s.get_graph();
    std::cout << s.find_path(start_node, finish_node, graph) << '\n';

    return 0;
}
